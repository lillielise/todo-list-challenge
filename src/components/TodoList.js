import "../styles/TodoList.css";
import { FaTrash, FaExclamationCircle } from "react-icons/fa";

const TodoList = ({
  todoListItems,
  handleDeleteItem,
  handlePrioritizeItem,
}) => {
  const renderTodoListItems = () => {
    return todoListItems.map((item) => (
      <li className="list-item" key={`todo-${item.id}`}>
        <p>{item.name}</p>
        {item.priority && <FaExclamationCircle color="#0074dd" />}
        <div className="list-item-icons">
          {!item.priority && (
            <FaExclamationCircle
              onClick={() => handlePrioritizeItem(item.id)}
            />
          )}
          <FaTrash onClick={() => handleDeleteItem(item.id)} />
        </div>
      </li>
    ));
  };
  return <ul className="list-items-container">{renderTodoListItems()}</ul>;
};

export default TodoList;
