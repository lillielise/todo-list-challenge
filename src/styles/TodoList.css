.list-items-container {
  height: 100%;
  overflow: scroll;
}

.list-item {
  padding: 1em;
  border-bottom: 1px solid #181a18;
  margin: 1em;
  display: flex;
  height: 15px;
  justify-content: space-between;
  overflow: auto;
}

.list-item-icons {
  display: flex;
  min-width: 40px;
}

.list-item-icons-hidden {
  display: none;
}

.list-item:hover .list-item-icons-hidden {
  display: block;
  cursor: pointer;
}

.list-item-icons *:not(:last-child) {
  margin-right: 0.5em;
}

.list-item-icons {
  margin-left: 0.5em;
}
