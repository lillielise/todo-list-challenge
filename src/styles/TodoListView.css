.todo-list-container {
  background-color: #f8f8ff;
  border-radius: 3px;
  box-shadow: 0 0 4px 0 #d3d3d3;
  width: 35vw;
  height: 80vh;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

.todo-list-header h1 {
  font-size: 2em;
  padding: 1em;
}
