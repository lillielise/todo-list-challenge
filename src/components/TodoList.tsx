import React from "react";
import "../styles/TodoList.css";
import { FaTrash, FaExclamationCircle } from "react-icons/fa";

type Props = {
  // Generics - describe what an array contains
  todoListItems: Array<Items>;
  handleDeleteItem: Function;
  handlePrioritizeItem: Function;
};

interface Items {
  id: number;
  name: string;
  priority: boolean;
}

const TodoList = ({
  todoListItems,
  handleDeleteItem,
  handlePrioritizeItem,
}: Props) => {
  const renderTodoListItems = () => {
    return todoListItems.map((item) => (
      <li className="list-item" key={`todo-${item.id}`}>
        <p>{item.name}</p>
        <div className="list-item-icons">
          {item.priority && <FaExclamationCircle color="#0074dd" />}
          <div className="list-item-icons-hidden">
            {!item.priority && (
              <FaExclamationCircle
                onClick={() => handlePrioritizeItem(item.id)}
              />
            )}
            <FaTrash onClick={() => handleDeleteItem(item.id)} />
          </div>
        </div>
      </li>
    ));
  };
  return <ul className="list-items-container">{renderTodoListItems()}</ul>;
};

export default TodoList;
