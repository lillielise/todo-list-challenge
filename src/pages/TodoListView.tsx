import React from "react";
import "../styles/TodoListView.css";
import { useState } from "react";
import TodoList from "../components/TodoList";
import AddTodoForm from "../components/AddTodoForm";

// Interface is exclusively used to represent the shape of an object data structure
// One major difference between type aliases vs interfaces is that interfaces are open and type aliases are closed. Meaning you can extend an interface by declaring it a second time
interface Items {
  id: number;
  name: string;
  priority: boolean;
}

const TodoListView = () => {
  const [todoListItems, setTodoListItems] = useState<Array<Items>>([]);

  const handleAddItem = (formData?: any, actions?: any): void => {
    const lastId = todoListItems.length
      ? todoListItems[todoListItems.length - 1].id
      : 0;
    setTodoListItems([
      ...todoListItems,
      { id: lastId + 1, name: formData.name, priority: false },
    ]);
    actions.resetForm();
  };

  const handleDeleteItem = (id: Number) => {
    setTodoListItems(todoListItems.filter((item) => item.id !== id));
  };

  const handlePrioritizeItem = (id: Number) => {
    // update item priority and sort todo list items - sort by id (order of insertion) if priority is the same and sort by priorty if they are not equal
    setTodoListItems(
      todoListItems
        .map((item) => (item.id === id ? { ...item, priority: true } : item))
        .sort((a, b): any => {
          if (a.priority === b.priority) {
            return a.id - b.id;
          } else {
            if (a.priority) {
              return -1;
            } else if (b.priority) {
              return 1;
            }
          }
          return null;
        })
    );
  };

  return (
    <div className="todo-list-container">
      <header className="todo-list-header">
        <h1>Todo List</h1>
      </header>
      <AddTodoForm handleAddItem={handleAddItem} />
      <TodoList
        handlePrioritizeItem={handlePrioritizeItem}
        handleDeleteItem={handleDeleteItem}
        todoListItems={todoListItems}
      />
    </div>
  );
};

export default TodoListView;
