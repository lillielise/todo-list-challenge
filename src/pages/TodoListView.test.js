import { shallow, mount } from "enzyme";
import TodoListView from "./TodoListView";
import TodoList from "../components/TodoList";
import "../setupTests";
import toJson from "enzyme-to-json";

it("renders without crashing", () => {
  // shallow used to render a single component - it doesn't render child components
  shallow(<TodoListView />);
});

it("renders header", () => {
  const wrapper = shallow(<TodoListView />);
  const welcome = <h1>Todo List</h1>;
  expect(wrapper.contains(welcome)).toEqual(true);
});

it("renders correctly", () => {
  const tree = shallow(<TodoListView />);
  // if we change the component then we need to update the snapshop to match - otherwise the test will fail
  expect(toJson(tree)).toMatchSnapshot();
});

const todoListItems = [{ id: 1, name: "walk dog", priority: true }];

describe("checking that view accepts todoListItems props", () => {
  it("accepts todoListItems props", () => {
    // mount is used  to render the full DOM, including child components of the parent running the tests
    const wrapper = mount(<TodoList todoListItems={todoListItems} />);
    expect(wrapper.props().todoListItems).toEqual(todoListItems);
  });
});

// will only work on class components
// it("renders correctly with no error message", () => {
//   const wrapper = mount(<TodoListView />);
//   expect(wrapper.state("error")).toEqual(null);
// });
