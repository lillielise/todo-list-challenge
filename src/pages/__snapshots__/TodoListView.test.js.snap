// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="todo-list-container"
>
  <header
    className="todo-list-header"
  >
    <h1>
      Todo List
    </h1>
  </header>
  <AddTodoForm
    handleAddItem={[Function]}
  />
  <TodoList
    handleDeleteItem={[Function]}
    handlePrioritizeItem={[Function]}
    todoListItems={Array []}
  />
</div>
`;

exports[`rendrs correctly 1`] = `
<div
  className="todo-list-container"
>
  <header
    className="todo-list-header"
  >
    <h1>
      Todo List
    </h1>
  </header>
  <AddTodoForm
    handleAddItem={[Function]}
  />
  <TodoList
    handleDeleteItem={[Function]}
    handlePrioritizeItem={[Function]}
    todoListItems={Array []}
  />
</div>
`;
